### PR TITLE
Splitting functional tests as cloud and non-cloud

### LINKED ISSUE #7624: github workflow changes to not require approval for non-cloud tests
### Area for Improvement

Currently, all functional tests require approval. The tests have already been separated out as cloud and non-cloud. This issue is to make github workflow/makefile changes so that no approval is required to run the non-cloud tests

### Observed behavior

_No response_

### Desired behavior

No approval is required to run noncloud tests
Approval is still required to run cloud tests

### Proposed Fix

_No response_

### rad Version

0.32.0

### Operating system

_No response_

### Additional context

_No response_

### Would you like to support us?

- [ ] Yes, I would like to support you

[AB#12400](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/12400)

### PR DESCRIPTION
**Pull request by ytimocin**
# Description

The goal of this PR is to create another workflow to run functional tests that don't use cloud resources and also remove the non-cloud tests from the existing functional test workflow that requires approval. In brief:

1. Workflow: **functional-test-cloud** (This requires an approval)
2. Workflow: **functional-test-noncloud** (This doesn't require an approval and it should be enabled on forks)

**RP tests that use cloud resources:**
* Core
* DataStores
* UCP

**RP tests that use non-cloud resources:**
* CLI
* Core
* Dapr
* DataStores
* Kubernetes
* Messaging
* Samples
* UCP

## Type of change
- This pull request adds or changes features of Radius and has an approved issue (issue link required).
Fixes: #7624

### CODE REVIEW

File: .github/workflows/functional-test-noncloud.yaml [hunk 1]
```diff
@@ -0,0 +1,697 @@
...
+          append: true
+          message: |
+            :x: Test tool installation for ${{ matrix.name }} failed. Please check [the logs](${{ env.ACTION_LINK }}) for more details
+      - name: Install Radius
+        run: |
+          export PATH=$GITHUB_WORKSPACE/bin:$PATH
+          which rad || { echo "cannot find rad"; exit 1; }
+
+          echo "*** Installing Radius to Kubernetes ***"
+          rad install kubernetes \
```
Comment (ytimocin) [inline .github/workflows/functional-test-noncloud.yaml:486]: Getting this error: `*** Installing Radius to Kubernetes ***
Installing Radius version afefa5b to namespace: radius-system...
Error: failed to run Radius Helm install, err: 
cannot re-use a name that is still in use
Helm output:
getting history for release radius
creating 1 resource(s)
creating 1 resource(s)
creating 1 resource(s)
beginning wait for 3 resources with timeout of 1m0s`.

Link to an example run: https://github.com/radius-project/radius/actions/runs/9659072267/job/26641840926.

Comment (ytimocin) [reply .github/workflows/functional-test-noncloud.yaml:486]: Added `--reinstall` flag and retesting: https://github.com/radius-project/radius/actions/runs/9662144721.

Comment (ytimocin) [reply .github/workflows/functional-test-noncloud.yaml:486]: Getting the same error.

Comment (ytimocin) [reply .github/workflows/functional-test-noncloud.yaml:500]: Removing `--reinstall`.

File: .github/workflows/functional-test-noncloud.yaml [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+      BICEP_RECIPE_TAG_VERSION: ${{ needs.build.outputs.REL_VERSION }}
+      DE_IMAGE: ${{ needs.build.outputs.DE_IMAGE }}
+      DE_TAG: ${{ needs.build.outputs.DE_TAG }}
+    steps:
+      - name: Get GitHub app token
+        uses: tibdex/github-app-token@v2
+        id: get_installation_token
+        with:
+          app_id: ${{ env.FUNCTIONAL_TEST_APP_ID }}
+          private_key: ${{ secrets.FUNCTIONAL_TEST_APP_PRIVATE_KEY }}
```
Comment (willdavsmith) [inline .github/workflows/functional-test-noncloud.yaml:343]: is there any risk of exposing secrets since these can run without approval?

File: test/functional-portable/datastoresrp/noncloud/resources/mongodb_test.go [hunk 3]
```diff
@@ -70,49 +70,3 @@ func Test_MongoDB_Manual(t *testing.T) {
 
 	test.Test(t)
 }
-
-// Test_MongoDB_Recipe validates:
-// the creation of a mongoDB from a recipe that uses an Azure resource
-func Test_MongoDB_Recipe(t *testing.T) {
```
Comment (ytimocin) [inline test/functional-portable/datastoresrp/noncloud/resources/mongodb_test.go:76]: Moved this test to the **cloud** package.

File: test/functional-portable/datastoresrp/noncloud/resources/redis_test.go [hunk 4]
```diff
@@ -72,75 +69,3 @@ func Test_Redis_Manual(t *testing.T) {
 
 	test.Test(t)
 }
-
-func Test_Redis_Recipe(t *testing.T) {
```
Comment (ytimocin) [inline test/functional-portable/datastoresrp/noncloud/resources/redis_test.go:76]: Moved these tests to the **cloud** package.

File: test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-mongodb-recipe-context.bicep
Comment (ytimocin) [inline test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-mongodb-recipe-context.bicep:1]: This wasn't being used anywhere.

Comment (ytimocin) [reply test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-mongodb-recipe-context.bicep:1]: Or maybe it was 🤨

File: test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-redis-manual.bicep
Comment (ytimocin) [inline test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-redis-manual.bicep:1]: Refactor was necessary.

File: test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-simulatedenv-recipe.bicep
Comment (ytimocin) [inline test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-simulatedenv-recipe.bicep:1]: Refactor was necessary.

File: test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-sqldb-manual.bicep
Comment (ytimocin) [inline test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-sqldb-manual.bicep:1]: Refactor was necessary.

File: test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-rs-mongodb-manual.bicep [hunk 5]
```diff
@@ -35,8 +35,6 @@ resource webapp 'Applications.Core/containers@2023-10-01-preview' = {
   }
 }
 
-
-// https://hub.docker.com/_/mongo/
```
Comment (ytimocin) [inline test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-rs-mongodb-manual.bicep:39]: Deleted this comment as I think it is unnecessary.

File: .github/workflows/functional-test.yaml [hunk 6]
```diff
@@ -33,31 +34,31 @@ on:
...
 
 env:
   # Go version
-  GOVER: '1.22.2'
+  GOVER: "1.22.2"
   GOPROXY: https://proxy.golang.org
 
   # gotestsum version - see: https://github.com/gotestyourself/gotestsum
-  GOTESTSUM_VER: 1.10.0
+  GOTESTSUM_VER: 1.12.0
```
Comment (ytimocin) [inline .github/workflows/functional-test.yaml:47]: It's been more than a year since the 1.10.0 version was released so I updated it: https://github.com/gotestyourself/gotestsum/releases.

File: .github/workflows/functional-test-noncloud.yaml [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+          echo "RAD_CLI_ARTIFACT_NAME=rad_cli_linux_amd64" >> $GITHUB_OUTPUT
+          echo "PR_NUMBER=${{ env.PR_NUMBER }}" >> $GITHUB_OUTPUT
+          echo "DE_IMAGE=${{ env.DE_IMAGE }}" >> $GITHUB_OUTPUT
+          echo "DE_TAG=${{ env.DE_TAG }}" >> $GITHUB_OUTPUT
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
```
Comment (ytimocin) [inline .github/workflows/functional-test-noncloud.yaml:168]: If this is being run from a fork, then logging into GHCR should be with that user's credentials.

File: .github/workflows/functional-test-noncloud.yaml [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+      - name: Build and Push container images
+        run: |
+          make build && make docker-build && make docker-push
```
Comment (ytimocin) [inline .github/workflows/functional-test-noncloud.yaml:171]: If this is from a fork, should these be published to the user's packages?

File: .github/actions/process-test-results/action.yaml
Comment (ytimocin) [inline .github/actions/process-test-results/action.yaml:1]: Reformatting all the files that I touch during the development of this PR.

File: .github/workflows/functional-test.yaml [hunk 9]
```diff
@@ -14,12 +14,13 @@
 # limitations under the License.
 # ------------------------------------------------------------
 
-name: Functional tests
+name: Functional Tests (with Cloud Resources)
```
Comment (lakshmimsft) [inline .github/workflows/functional-test.yaml:17]: Could we rename this file to functional-test-cloud.yaml?

File: test/functional-portable/datastoresrp/cloud/resources/redis_test.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+	"testing"
+
+	"github.com/radius-project/radius/test/rp"
+	"github.com/radius-project/radius/test/step"
+	"github.com/radius-project/radius/test/testutil"
+	"github.com/radius-project/radius/test/validation"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_Redis_Recipe(t *testing.T) {
```
Comment (lakshmimsft) [inline test/functional-portable/datastoresrp/cloud/resources/redis_test.go:31]: I'm missing something. why is this moved to cloud resources tests?

Comment (ytimocin) [reply test/functional-portable/datastoresrp/cloud/resources/redis_test.go:31]: It is using an Azure provider in the environment. Please see: https://github.com/radius-project/radius/blob/b961aa02f67375eb5e809559cfad77f8d5509c6f/test/functional-portable/datastoresrp/cloud/resources/testdata/datastoresrp-resources-redis-recipe.bicep.

**Review by lakshmimsft — DISMISSED** [review]

File: .github/workflows/functional-test-noncloud.yaml [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+  # KinD cluster version
+  KIND_VER: "v0.20.0"
+  # Dapr version
+  DAPR_VER: "1.12.0"
+  # Dapr dashboard version
+  DAPR_DASHBOARD_VER: "0.14.0"
+  # Kubectl version
+  KUBECTL_VER: "v1.25.0"
+  # Container registry for storing container images
+  CONTAINER_REGISTRY: ghcr.io/${{ github.repository_owner }}/dev
```
Comment (ytimocin) [inline .github/workflows/functional-test-noncloud.yaml:62]: We can run a registry in the Kubernetes cluster.

Comment (ytimocin) [reply .github/workflows/functional-test-noncloud.yaml:62]: Add `radius` in the URL.

**Comment by lakshmimsft** [comment]
The Approval workflow today triggers ensures all tests are run. Since we're splitting functional-tests into cloud and non-cloud, will that be updated? to run only cloud tests or no?
How do we ensure in the future that both cloud and non-cloud tests have be run?

**Comment by ytimocin** [comment]
> The Approval workflow today triggers ensures all tests are run. Since we're splitting functional-tests into cloud and non-cloud, will that be updated? to run only cloud tests or no? How do we ensure in the future that both cloud and non-cloud tests have be run?

When someone creates a PR the functional tests with non-cloud resources will run by default. The other one, functional tests with cloud resources, will need an approval like it does today. Basically, all the tests will have run. Here is where the one with the approval is: https://github.com/radius-project/radius/pull/7716/files#diff-a5e3050f5c1a812d9565ba0c87fb5028ad0a004c2d966e71d67208b4b33042c4L327-R332.

File: build/recipes.mk [hunk 12]
```diff
@@ -15,7 +15,6 @@
 # ------------------------------------------------------------
 
 BICEP_RECIPE_TAG_VERSION?=latest
-RAD_BICEP_PATH?=$${HOME}/.rad/bin
```
Comment (ytimocin) [inline build/recipes.mk:18]: This is not being used anywhere in the code. I did a search in the whole project. Please correct me if I am wrong.

File: test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-mongodb-recipe-context.bicep
Comment (ytimocin) [inline test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-mongodb-recipe-context.bicep:1]: This wasn't being used.

File: .github/workflows/functional-test-noncloud.yaml [hunk 13]
```diff
@@ -0,0 +1,445 @@
...
+  build:
+    name: Build Radius for test
+    runs-on: ubuntu-latest
+    env:
+      DE_IMAGE: "ghcr.io/radius-project/deployment-engine"
+      DE_TAG: "latest"
+    outputs:
+      REL_VERSION: ${{ steps.gen-id.outputs.REL_VERSION }}
+      DE_IMAGE: ${{ steps.gen-id.outputs.DE_IMAGE }}
+      DE_TAG: ${{ steps.gen-id.outputs.DE_TAG }}
```
Comment (rynowak) [inline .github/workflows/functional-test-noncloud.yaml:95]: Where do these get set? `gen-id` runs as part of this job, so how can it provide the outputs?

File: .github/workflows/functional-test-noncloud.yaml [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+          echo "CHECKOUT_REPO=${{ github.repository }}" >> $GITHUB_ENV
+          echo "CHECKOUT_REF=refs/heads/main" >> $GITHUB_ENV
+
+      - name: Set up checkout target (pull_request)
+        if: github.event_name == 'pull_request'
+        run: |
+          echo "CHECKOUT_REPO=${{ github.repository }}" >> $GITHUB_ENV
+          echo "CHECKOUT_REF=${{ github.ref }}" >> $GITHUB_ENV
+          echo "PR_NUMBER=${{ github.event.pull_request.number }}" >> $GITHUB_ENV
+
```
Comment (rynowak) [inline .github/workflows/functional-test-noncloud.yaml:174]: Don't we also need to setup the checkout target for a workflow dispatch here?

File: .github/workflows/functional-test-noncloud.yaml [hunk 15]
```diff
@@ -0,0 +1,445 @@
...
+      - name: Use custom actions
```
Comment (rynowak) [inline .github/workflows/functional-test-noncloud.yaml:175]: This doesn't sound right 😆

File: .github/workflows/functional-test-noncloud.yaml [hunk 16]
```diff
@@ -0,0 +1,445 @@
...
+          registry-server: ${{ env.LOCAL_REGISTRY_SERVER }}
+          registry-port: ${{ env.LOCAL_REGISTRY_PORT }}
+
+      - name: Build and Push container images
+        run: |
+          make build && make docker-build && make docker-push
+        env:
+          DOCKER_REGISTRY: "${{ env.LOCAL_REGISTRY_SERVER }}:${{ env.LOCAL_REGISTRY_PORT }}"
+          DOCKER_TAG_VERSION: ${{ env.REL_VERSION }}
+
```
Comment (rynowak) [inline .github/workflows/functional-test-noncloud.yaml:223]: Very cool to see this come together!

File: .github/workflows/functional-test-noncloud.yaml [hunk 17]
```diff
@@ -0,0 +1,445 @@
...
+        if: failure() && github.repository == 'radius-project/radius'
+        with:
+          test_group_name: "Functional Tests - ${{ matrix.name }}"
+          artifact_name: "functional_test_results_${{ matrix.name }}"
+          result_directory: "dist/functional_test/"
+
+      - uses: azure/setup-kubectl@v4
+        if: always()
+        with:
+          version: ${{ env.KUBECTL_VER }}
```
Comment (rynowak) [inline .github/workflows/functional-test-noncloud.yaml:336]: Do we need this here? I think the scripts already called into kubectl. IMO it's better to install dependencies early in the workflow instead of near the end.

File: build/recipes.mk (hunk 12 above)
Comment (rynowak) [reply build/recipes.mk:18]: If it's unused then it's good for us to delete !

File: test/functional-portable/cli/noncloud/cli_test.go [hunk 18]
```diff
@@ -68,7 +68,9 @@ func verifyRecipeCLI(ctx context.Context, t *testing.T, test rp.RPTest) {
 	version := strings.TrimPrefix(testutil.GetBicepRecipeVersion(), "version=")
 	resourceType := "Applications.Datastores/redisCaches"
 	file := "../../../testrecipes/test-bicep-recipes/corerp-redis-recipe.bicep"
-	target := fmt.Sprintf("br:ghcr.io/radius-project/dev/test-bicep-recipes/redis-recipe:%s", generateUniqueTag())
+
+	// The target is a local registry, so we can use localhost.
+	target := fmt.Sprintf("br:localhost:5000/dev/test-bicep-recipes/redis-recipe:%s", generateUniqueTag())
```
Comment (rynowak) [inline test/functional-portable/cli/noncloud/cli_test.go:73]: Should we pass this in via an env-var since this requires additional setup? Then we could skip if the env-var is not set.

It's best if the functional tests "just work" without additional steps like a local registry.

File: test/functional-portable/datastoresrp/cloud/resources/redis_test.go
[the remark this answers is not in the document]
Comment (rynowak) [reply test/functional-portable/datastoresrp/cloud/resources/redis_test.go:1]: Do we have coverage of these resources without cloud resources? I thought all of the datastores RP code was tested without a cloud dependency..

File: test/functional-portable/datastoresrp/cloud/resources/redis_test.go (hunk 10 above)
Comment (rynowak) [reply test/functional-portable/datastoresrp/cloud/resources/redis_test.go:31]: Why though? Does it actually require an Azure resource? 

We should have a justification when tests require the cloud.

**Review by rynowak — CHANGES_REQUESTED** [review]
Great to see the progress on this!

I think my only major concern is about the tests that are being moved to the cloud package. Can we huddle quickly and figure out why this is needed?

We should use cloud resources iff we need to test cloud features.

File: .github/actions/create-kind-cluster/action.yaml [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+        for node in $(kind get nodes); do
+          # LOCALHOST_DIR
+          docker exec "${node}" mkdir -p "${LOCALHOST_DIR}"
+          cat <<EOF | docker exec -i "${node}" cp /dev/stdin "${LOCALHOST_DIR}/hosts.toml"
+        [host."http://${{ inputs.registry-name }}:${{ inputs.registry-port }}"]
+          capabilities = ["pull", "resolve", "push"]
+          skip_verify = true
+        EOF
+
+          # RADIUS_DIR
```
[the remark this answers is not in the document]
Comment (ytimocin) [reply .github/actions/create-kind-cluster/action.yaml]: Removed

File: .github/actions/create-kind-cluster/action.yaml [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+        cat <<EOF | kind create cluster --config=-
+        kind: Cluster
+        apiVersion: kind.x-k8s.io/v1alpha4
+        EOF
+
+    - name: Create a KinD cluster with a local registry
+      if: ${{ inputs.with-local-registry == 'true' }}
+      shell: bash
+      run: |
+        cat <<EOF | kind create cluster --config=-
```
[the remark this answers is not in the document]
Comment (ytimocin) [reply .github/actions/create-kind-cluster/action.yaml:49]: Yes, and added the reference.

File: .github/actions/create-kind-cluster/action.yaml [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+    required: false
+    default: "radius-registry"
+  registry-server:
+    description: "The server name for the local registry"
+    required: false
+    default: "localhost"
+  registry-port:
+    description: "The port for the local registry"
+    required: false
+    default: "5000"
```
[the remark this answers is not in the document]
Comment (ytimocin) [reply .github/actions/create-kind-cluster/action.yaml:24]: We don't need any output for now. The callee has context of the URL and the registry name because they can be changed by the callee.

File: .github/actions/create-local-registry/action.yaml [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+    default: "radius-registry"
+  registry-server:
+    description: "The server name for the local registry"
+    required: false
+    default: "localhost"
+  registry-port:
+    description: "The port for the local registry"
+    required: false
+    default: "5000"
+runs:
```
[the remark this answers is not in the document]
Comment (ytimocin) [reply .github/actions/create-local-registry/action.yaml:23]: Same comment as above.

File: .github/workflows/functional-test-noncloud.yaml (hunk 17 above)
Comment (ytimocin) [reply .github/workflows/functional-test-noncloud.yaml]: I removed this step from both workflows.

**Comment by ytimocin** [comment]
https://github.com/radius-project/radius/pull/7716#discussion_r1683629719 There were some bicep files that used Azure provider for the environment and that is why I had moved those tests to the cloud package in the datastoresrp functional test. Then I checked and saw that we don't need the provider. I removed those parts and moved all those tests back to the noncloud package. cc/ @rynowak

File: .github/actions/create-kind-cluster/action.yaml [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+      shell: bash
+      run: |
+        cat <<EOF | kind create cluster --config=-
+        kind: Cluster
+        apiVersion: kind.x-k8s.io/v1alpha4
+        nodes:
+        - role: control-plane
+          extraMounts:
+            - containerPath: "/etc/containerd/certs.d/${{ inputs.registry-name }}"
+              hostPath: "$TEMP_CERT_DIR/certs/${{ inputs.registry-server }}"
```
[the remark this answers is not in the document]
Comment (rynowak) [reply .github/actions/create-kind-cluster/action.yaml:56]: Having this environment variable set is a prerequisite for using this action right? Wouldn't that be better as a parameter since it's required?

File: test/functional-portable/datastoresrp/noncloud/resources/mongodb_test.go (hunk 3 above)
Comment (rynowak) [reply test/functional-portable/datastoresrp/noncloud/resources/mongodb_test.go:76]: Thank you! 🙇

File: .github/actions/create-kind-cluster/action.yaml (hunk 23 above)
[the remark this answers is not in the document]
Comment (ytimocin) [reply .github/actions/create-kind-cluster/action.yaml:56]: Oh, I see now. I will update this.

File: test/testutil/testutil.go [hunk 24]
```diff
@@ -335,3 +335,12 @@ func GetCreationTimestamp() string {
 func GenerateS3BucketName() string {
 	return "radiusfunctionaltestbucket-" + uuid.New().String()
 }
+
+func GetDockerRegistryName() string {
```
Comment (rynowak) [inline test/testutil/testutil.go:339]: How is this different from `DOCKER_REGISTRY`? Looks like they are two different settings with the same default value. Maybe a comment would help?

Comment (ytimocin) [reply test/testutil/testutil.go:342]: I added a comment. In the case of using a local registry, DOCKER_REGISTRY will return the port too, so I added this function that returns the name of the registry without a port if present.

File: .github/actions/create-kind-cluster/action.yaml (hunk 23 above)
[the remark this answers is not in the document]
Comment (rynowak) [reply .github/actions/create-kind-cluster/action.yaml]: This is still using the environment variable. Can you update it please?

File: .github/actions/create-kind-cluster/action.yaml [hunk 25]
```diff
@@ -0,0 +1,148 @@
+name: "Create a KinD cluster"
+description: |
+  Create a KinD cluster.
+inputs:
+  secure:
+    description: "Whether the KinD cluster should be created with a secure local registry configuration"
+    required: false
+    default: "false"
+  temp-cert-dir:
```
Comment (rynowak) [inline .github/actions/create-kind-cluster/action.yaml:9]: suggestion: `cert-dir`. Saying "temp" doesn't really make sense as a parameter name. I think it's a translation from the original code.

**Review by rynowak — DISMISSED** [review]

**Review by sylvainsf — APPROVED** [review]
Looks great!
